ppafm/io.py

```diff
@@ -415,9 +416,68 @@ def loadGeometry(fname=None,format=None,params=None):
         if qs is None:
             qs = np.zeros(len(Zs))
         atoms = [list(Zs), list(xyzs[:, 0]), list(xyzs[:, 1]), list(xyzs[:, 2]), list(qs)]
-        nDim = params['gridN'].copy()
+        nDim = params["gridN"].copy()
 
-    return atoms,nDim,lvec
+    # Make sure lvec is a 4x3 array. Use default grid parameters if needed
+    if (lvec is None) or (len(lvec) == 0):
+        lvec = np.zeros((4, 3))
+        lvec[1, :] = params["gridA"].copy()
+        lvec[2, :] = params["gridB"].copy()
+        lvec[3, :] = params["gridC"].copy()
+    else:
+        if len(lvec) == 1:
+            lvec.append(params["gridA"].copy())
+        if len(lvec) == 2:
+            lvec.append(params["gridB"].copy())
+        if len(lvec) == 3:
+            lvec.append(params["gridC"].copy())
+    lvec = np.array(lvec)
+
+    # Shift scanning coordinates to such that actually make sense (being directly comparable to coordinates of atoms)
+    params["scanMin"][0] += params["r0Probe"][0]
+    params["scanMin"][1] += params["r0Probe"][1]
+    params["scanMin"][2] -= params["r0Probe"][2]
+    params["scanMax"][0] += params["r0Probe"][0]
+    params["scanMax"][1] += params["r0Probe"][1]
+    params["scanMax"][2] -= params["r0Probe"][2]
+
+    # Generate automatic lattice vectors and grid dimensions if needed
+    pad = 3.0
+    default_grid_step = 0.1
+    for i in range(3):
+        # Zero lattice vector is considered undefined and triggers creation of an automatic one.
+        # The automatic generated lattice vector should enclose the whole area filled with atoms as well as the whole scanning area, plus the default padding.
+        if np.allclose(lvec[i + 1, :], 0):
+            if params["PBC"]:
+                lvec[i + 1, i] = (
+                    max(max(atoms[i + 1]), params["scanMax"][i])
+                    - min(min(atoms[i + 1]), params["scanMin"][i])
+                    + 2 * pad
+                )
```

Comment:
Do we really need to consider the atom positions here? It should be enough to just have enough padding around the scan region so that the probe particle stays inside the FF grid during the scan. Sometimes you just want a smaller scan of a bigger structure.

Reply:
Yes, you are right. I will redo it.